packages/material-ui-system/src/index.d.ts

```diff
@@ -150,18 +150,18 @@ export const spacing: SimpleStyleFunction<
 export type SpacingProps = PropsFor<typeof spacing>;
 
 // style.js
-export interface StyleOptions<PropKey, Theme extends object> {
+export interface StyleOptions<PropKey, Theme extends object, PropType> {
   cssProperty?: PropKey | keyof CSS.Properties | false;
   prop: PropKey;
   /**
    * dot access in `Theme`
    */
   themeKey?: string;
-  transform?: (cssValue: unknown) => number | string;
+  transform?: (cssValue: PropType) => number | string | { [k: string]: any };
 }
-export function style<PropKey extends string, Theme extends object>(
-  options: StyleOptions<PropKey, Theme>,
-): StyleFunction<{ [K in PropKey]: unknown } & { theme: Theme }>;
+export function style<PropKey extends string, Theme extends object, PropType>(
+  options: StyleOptions<PropKey, Theme, PropType>,
+): StyleFunction<{ [K in PropKey]?: PropType } & { theme?: Theme }>;
```

Comment:
Isn't the actual issue with `styled` that doesn't correctly make the `theme` prop optional in decorated components? 

I think this should stay as is to make it obvious that it's meant to be used with `styled` or the `theme` has to be injected into props by other means.

Reply:
That was the decision point I was talking about before. Currently the built-in style functions like palette which are created by the same `style`-function do not expose theme as required, even though they do require it too. Do we want to:

1. stay in line with the built-in function typings and not require theme
2. adjust the built-in function typings to require theme & alter `styled` typings
3. leave built-in types as-is (theme not required), leave style as-is (theme required) and adjust `styled` typings?

Reply:
just give me a 1, 2, 3 or possibly 4 I did not spot and I make the changes.

Reply:
For 2 & 3 change in `styled` would look something like this:

```ts
export type ComponentCreator<Component extends React.ElementType> = <Theme, Props extends {} = any>(
  styles:
    | CreateCSSProperties<Props>
    | (({ theme, ...props }: { theme: Theme } & Props) => CreateCSSProperties<Props>),
  options?: WithStylesOptions<Theme>,
) => React.ComponentType<
  Omit<
    JSX.LibraryManagedAttributes<Component, React.ComponentProps<Component>>,
    'classes' | 'className'
  > &
    StyledComponentProps<'root'> & { className?: string } & CoerceEmptyInterface<
      Props extends { theme: Theme } ? Omit<Props, 'theme'> & { theme?: Props['theme'] } : Props
    >
>;
```

And theme would be non-nullable on `StyleFunction`.

Reply:
I think the typings for the built-ins are just wrong. In a perfect world those would just reference the same type `style` returns. The actual issue seems to be `styled` that injects the theme. I wasn't aware of this behavior nor did I find any runtime test for it.

Reply:
I have the direction for next changes then thanks! :) I totally agree with you